disttask/framework/dispatcher/dispatcher.go

```diff
@@ -418,34 +417,26 @@ func (d *dispatcher) processNormalFlow(gTask *proto.Task) (err error) {
 	}
 
 	// Generate all available TiDB nodes for this global tasks.
-	serverNodes, err1 := GenerateSchedulerNodes(d.ctx)
+	serverNodes, err1 := handle.GetEligibleInstances(d.ctx, gTask)
 	if err1 != nil {
 		return err1
 	}
+	if len(serverNodes) == 0 {
+		return errors.New("no available TiDB node")
+	}
 	subTasks := make([]*proto.Subtask, 0, len(metas))
 	for i, meta := range metas {
-		instanceID, err := GetEligibleInstance(serverNodes, i)
-		if err != nil {
-			logutil.BgLogger().Warn("get a eligible instance failed", zap.Int64("gTask ID", gTask.ID), zap.Error(err))
-			return err
-		}
+		instanceID := GetInstanceForSubtask(serverNodes, i)
 		subTasks = append(subTasks, proto.NewSubtask(gTask.ID, gTask.Type, instanceID, meta))
 	}
 	return d.updateTask(gTask, gTask.State, subTasks, retrySQLTimes)
 }
 
-// GetEligibleInstance gets an eligible instance.
-func GetEligibleInstance(serverNodes []*infosync.ServerInfo, pos int) (string, error) {
-	if pos >= len(serverNodes) && pos < 0 {
-		errMsg := fmt.Sprintf("available TiDB nodes range is 0 to %d, but request position: %d", len(serverNodes)-1, pos)
-		return "", errors.New(errMsg)
-	}
-	if len(serverNodes) == 0 {
```

Comment:
Why remove these check? My be once using this function，and len is 0.

Reply:
previous `errMsg := fmt.Sprintf("available` is dead code, never reaches

below one move it to caller site `	if len(serverNodes) == 0 {
		return errors.New("no available TiDB node")`

Reply:
It think this check put this function is more reasonable. After all, this function doesn't have to be called in this PR.

Reply:
https://github.com/pingcap/tidb/pull/43880/commits/01bba1351dec9bb90ef49d6a1c940507b160cf0f